pkg/api/sort.go

```diff
@@ -0,0 +1,388 @@
+	return s.Name
+}
+
+func (s udpServiceRepresentation) resourceType() string {
+	return s.Type
+}
+
+func (s udpServiceRepresentation) serversCount() int {
+	// TODO: maybe disable that data point altogether,
+	// if we can't/won't compute a fully correct (recursive) result.
+	// Or "redefine" it as only the top-level count?
+	// Note: The current algo is equivalent to the webui one.
+	if s.LoadBalancer == nil {
+		return 0
+	}
+
+	return len(s.LoadBalancer.Servers)
+}
+
+func (s udpServiceRepresentation) provider() string {
+	return s.Provider
+}
+
+func (s udpServiceRepresentation) status() string {
+	return s.Status
+}
+
+type orderedMiddleware interface {
+	orderedWithName
+
+	resourceType() string
+	provider() string
+	status() string
+}
+
+func sortMiddlewares[T orderedMiddleware](values url.Values, middlewares []T) {
+	sortBy := values.Get(sortByParam)
+
+	direction := values.Get(directionParam)
+	if direction == "" {
+		direction = ascendantSorting
+	}
+
+	switch sortBy {
+	case "name":
+		sortByName(direction, middlewares)
+
+	case "type":
+		sortMeta(direction, middlewares, func(i int) string { return middlewares[i].resourceType() })
+
+	case "provider":
+		sortMeta(direction, middlewares, func(i int) string { return middlewares[i].provider() })
+
+	case "status":
+		sortMeta(direction, middlewares, func(i int) string { return middlewares[i].status() })
+
+	default:
+		sortByName(direction, middlewares)
+	}
+}
+
+func (m middlewareRepresentation) name() string {
+	return m.Name
+}
+
+func (m middlewareRepresentation) resourceType() string {
+	return m.Type
+}
+
+func (m middlewareRepresentation) provider() string {
+	return m.Provider
+}
+
+func (m middlewareRepresentation) status() string {
+	return m.Status
+}
+
+func (m tcpMiddlewareRepresentation) name() string {
+	return m.Name
+}
+
+func (m tcpMiddlewareRepresentation) resourceType() string {
+	return m.Type
+}
+
+func (m tcpMiddlewareRepresentation) provider() string {
+	return m.Provider
+}
+
+func (m tcpMiddlewareRepresentation) status() string {
+	return m.Status
+}
+
+type orderedByName interface {
+	orderedWithName
+}
+
+func sortByName[T orderedByName](direction string, results []T) {
+	// Ascending
+	if direction == ascendantSorting {
+		sort.Slice(results, func(i, j int) bool {
+			return results[i].name() < results[j].name()
+		})
+
+		return
+	}
+
+	// Descending
+	sort.Slice(results, func(i, j int) bool {
+		return results[i].name() > results[j].name()
+	})
+}
+
+func sortMeta[T orderedWithName, U constraints.Ordered](direction string, results []T, fn func(int) U) {
+	// Ascending
+	if direction == ascendantSorting {
+		sort.Slice(results, func(i, j int) bool {
+			if fn(i) == fn(j) {
+				return results[i].name() < results[j].name()
+			}
+
+			return fn(i) < fn(j)
+		})
+
+		return
+	}
+
+	// Descending
+	sort.Slice(results, func(i, j int) bool {
+		if fn(i) == fn(j) {
+			return results[i].name() > results[j].name()
+		}
+		return fn(i) > fn(j)
+	})
```

Comment:
```suggestion
	sortByName(direction, results)

	// Ascending
	if direction == ascendantSorting {
		sort.Slice(results, func(i, j int) bool {
			return fn(i) < fn(j)
		})

		return
	}

	// Descending
	sort.Slice(results, func(i, j int) bool {
		return fn(i) > fn(j)
	})
```

This could be clearer, but at the expense of some perfs